File: 695-max-area-of-island/695-max-area-of-island.cpp
class Solution {
public:
    int max_area = 0;
    int solve(vector<vector<int>>& grid,vector<vector<int>>& v, int x, int y)
    {
        int m = grid.size();
        int n = grid[0].size();
        if (x < 0 || x >= m || y < 0 || y >= n || v[x][y] || !grid[x][y])
        {
            return 0;
        }
        
        v[x][y] = 1;
        
        return 1 + solve(grid, v, x + 1, y) + 
            solve(grid, v, x, y + 1) + 
            solve(grid, v, x - 1, y) + 
            solve(grid, v, x, y - 1);
    }
    
    int maxAreaOfIsland(vector<vector<int>>& grid) 
    {   
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<int>> v(m, vector<int>(n, 0));
        for (int i=0; i<m; i++)
        {
            for (int j=0; j<n; j++)
            {
                max_area = max(max_area, solve(grid, v, i, j));
            }
        }
        return max_area;
    }
};

File: 695-max-area-of-island/695-max-area-of-island.py
class Solution:
    def __init__(self):
        self.visited = []
        
    def solve(self, grid, rows, cols, x, y):
        if not ((0 <= x < rows) and (0 <= y < cols) and grid[x][y] and not self.visited[x][y]):
            return 0
        self.visited[x][y] = 1
        return 1 +  self.solve(grid, rows, cols, x + 1, y) + \
                    self.solve(grid, rows, cols, x, y + 1) + \
                    self.solve(grid, rows, cols, x - 1, y) + \
                    self.solve(grid, rows, cols, x, y - 1)
        
    
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        rows = len(grid)
        cols = len(grid[0])
        self.visited = [[0 for i in range(cols)] for j in range(rows)]
        max_area = 0
        for x in range(rows):
            for y in range(cols):
                if (grid[x][y] == 1 and self.visited[x][y] == 0):
                    max_area = max(max_area, self.solve(grid, rows, cols, x, y))
        return max_area
    
    
    